src/construct.ts

```diff
@@ -376,90 +318,54 @@ export class Node {
-        };
-        (construct as any).onSynthesize(ctx); // "as any" is needed because we want to keep "synthesize" protected
-      } finally {
-        node._unlock();
-      }
-    }
+  public addValidation(validation: IValidation) {
+    this._validations.push(validation);
   }
 
   /**
-   * Invokes "prepare" on all constructs (depth-first, post-order) in the tree under `node`.
-   */
-  public prepare() {
-    // Aspects are applied root to leaf
-    for (const construct of this.findAll(ConstructOrder.PREORDER)) {
-      Node.of(construct).invokeAspects();
-    }
-
-    // since constructs can be added to the tree during invokeAspects, call findAll() to recreate the list.
-    // use PREORDER.reverse() for backward compatability
-    for (const construct of this.findAll(ConstructOrder.PREORDER).reverse()) {
-      const cn = construct as any;
-      if ('onPrepare' in cn) {
-        if (typeof(cn.onPrepare) !== 'function') { throw new Error('expecting "onPrepare" to be a function'); }
-        cn.onPrepare();
+   * Validates this construct.
+   *
+   * Invokes the `validate()` method on all validations added through
+   * `addValidation()`.
+   *
+   * @returns an array of validation error messages associated with this
+   * construct.
+   */
+  public validate(): string[] {
+    const deprecated = ['validate', 'onValidate', 'synthesize', 'onSynthesize', 'prepare', 'onPrepare'];
+    for (const method of deprecated) {
+      if (typeof((this.host as any)[method]) === 'function') {
+        throw new Error(`the construct "${this.path}" has a "${method}()" method which is no longer supported. Use "construct.node.addValidation()" to add validations to a construct`);
       }
     }
-  }
-
-  /**
-   * Invokes "validate" on all constructs in the tree (depth-first, pre-order) and returns
-   * the list of all errors. An empty list indicates that there are no errors.
-   */
-  public validate() {
-    let errors = new Array<ValidationError>();
 
-    for (const child of this.children) {
-      errors = errors.concat(Node.of(child).validate());
+    const errors = new Array<string>();
+    for (const v of this._validations) {
+      errors.push(...v.validate());
     }
 
-    const localErrors: string[] = (this.host as any).onValidate(); // "as any" is needed because we want to keep "validate" protected
-    return errors.concat(localErrors.map(msg => ({ source: this.host, message: msg })));
+    return errors;
   }
 
   /**
    * Locks this construct from allowing more children to be added. After this
    * call, no more children can be added to this construct or to any children.
-   * @internal
    */
-  private _lock() {
+  public lock() {
     this._locked = true;
   }
 
   /**
-   * Unlocks this costruct and allows mutations (adding children).
-   * @internal
+   * Unlocks this construct and allows mutations (adding children).
    */
-  private _unlock() {
+  public unlock() {
     this._locked = false;
   }
```

Comment:
I agree with the use case for `lock()` but I don't see why we need `unlock()` yet. Once it's locked, it's locked for the rest of the process.

Reply:
Good point, I'll remove `unlock()` for now.

Reply:
Removed.